libs/wingsdk/src/target-sim/test-runner.inflight.ts

```diff
@@ -0,0 +1,59 @@
+import { ISimulatorResourceInstance } from "./resource";
+import { TestRunnerAttributes, TestRunnerSchema } from "./schema-resources";
+import { IFunctionClient, ITestRunnerClient, TestResult } from "../cloud";
+import { ISimulatorContext } from "../testing";
+
+export class TestRunnerClient
+  implements ITestRunnerClient, ISimulatorResourceInstance
+{
+  // A map from test paths to their corresponding function handles.
+  private readonly tests: Map<string, string>;
+  private readonly context: ISimulatorContext;
+
+  constructor(props: TestRunnerSchema["props"], context: ISimulatorContext) {
+    this.tests = new Map(Object.entries(props.tests));
+    this.context = context;
+  }
+
+  public async init(): Promise<TestRunnerAttributes> {
+    return {};
+  }
+
+  public async cleanup(): Promise<void> {
+    return;
+  }
+
+  public async listTests(): Promise<string[]> {
+    return Array.from(this.tests.keys());
+  }
+
+  public async runTest(path: string): Promise<TestResult> {
```

Comment:
Do we really need a different implementation for this between aws and the simulator? Intuitively the test runner should be able to simply use cloud APIs and be implemented once for all targets. No?

Reply:
Mmm... it sounds possible? Though I'm not sure I see a clear path to doing it right now, at least two issues come to mind:
- aws function client and simulator function client constructors expecting different kinds of values (one expects simulator handles and other simulator context, the other expects AWS arns)
- logic for obtaining logs/traces for the tests might be very different between the two targets

(Totally open to adding a base class in the future)

(Some WET code is also probably okay for now)

Reply:
Something feels off... Let me challenge you like this: what would it take to implement `TestRunner` in Wing (just using `cloud.Xxx` resources)?

For what it's worth, it's a good dogfooding exercise...

Reply:
@eladb I think to implement a `TestRunner` in Wing we are missing at least a few more capabilities, like construct tree inspection, logging, and an API for errors in the Wing language. Here's an idea of what I have sketched out:

```js
bring cloud;

struct TestResult {
  path: str;
  pass: bool;
  error: str?;
  logs: Array<str>;
}

resource TestRunner {
  fn: cloud.Function;
  init(fn: cloud.Function) {
    this.fn = fn;
    // TODO
    // this.fns = nodeof(this).find_all().filter(Test.is_test);
  }
  inflight run_test(): TestResult {
    let var pass = false;
    let var error: str?;
    try {
      this.fn.invoke("");
      pass = true;
    } catch e {
      error = e.message; // TODO: no error API in Wing yet
    }
    let logs = this.fn.logs(tail: 100); // TODO: no logs API in Wing yet
    return TestResult {
      path: nodeof(this.fn).path, // TODO: no `nodeof` API in Wing yet
      pass: true,
      error: error,
      logs: Array<str>[],
    };
  }
}
```

thoughts? Do you think any of these capabilities should be pulled forward into the current sprint?

Reply:
The original question was whether we can implement this once (in typescript) for all targets and avoid target specific code (and logic duplication).

Seems that for typescript the main missing thing is a way to collect logs, correct?

Let's make sure we have issues for all the missing pieces.

Reply:
https://github.com/winglang/wing/issues/1987